### ISSUE TITLE
windowManager.open() - classes Property?

### ISSUE THREAD

**Issue by manu37** [tinymce/tinymce#4761]
In version 4 there was a 'classes' for windowManager.open() to set a class on the dialog box. How do I achieve this in v5?

**Comment by Paril** [comment]
I was just running into the same issue. I now have no way to uniquely identify a dialog, or style them effectively. For instance, I have specific cells that need to be styled for a certain font (material icons) but all of the DOM and such is opaque to the programmer now.

**Comment by TheSpyder** [comment]
I've been thinking about this myself to make testing better. The effect of the current decision is we've had to rely on `[role="dialog"]` which isn't unique. This test, for example, will pass for any dialog that has buttons with the correct name when the intention is to test the link dialog.
https://github.com/tinymce/tinymce/blob/5.x/src/plugins/link/test/ts/module/TestLinkUi.ts#L24

I'm not sure whether we'd offer classes on every component as v4 did or just on the dialog itself, but I'll bring it up with the team.

**Comment by Paril** [comment]
It might be nice to have it at component level. There's currently no way to uniquely identify anything in tox containers. IDs generated by forms are random, it seems. How would we ever be able to style these things easily unless we were to make a separate component, which seems overkill for something like "this label needs to have a different font"

**Comment by TheSpyder** [comment]
That's the intent of the new UI, though. It's clean, consistent and accessible. We're always open to discussion of use cases we haven't considered, but right now for a user there would be no weird edge cases to learn. Dialogs all look and operate in a consistent way.

Styling _all_ labels is intended to be easy through the skin (there's an updated theme customisation tool in the works). Styling _one_ label is not.

**Comment by Paril** [comment]
If your label has more than a few words of text, it just flows off of the side of the dialog. For instance:

!()[https://i.imgur.com/M9WCssy.png]

I had to kind of do a workaround hack to change "white-space" to "normal". I don't really want to edit the skin JS or add a new control type just to make a label wrap onto more lines. This doesn't seem like an edge case to me, just that the dialogs were only designed with the built-in dialogs in mind and not taking what end-users are going to do with them into account. See my other ticket about editing the "charmap" plugin to add more tabs to it; there's no explicit height or anything in the plugin itself, but it still seems to snap to that height.

I love the idea of the new dialogs, don't get me wrong; flex is amazing, but it seems like it's being used entirely wrong here if elements aren't flexing or wrapping at all.

**Comment by TheSpyder** [comment]
This label wrap and the charmap size issue are definitely bugs. We made a mistake during development by not adding a reset style until after the dialogs were designed; the effect is certainly that some things now only work for the built-in dialogs but that is definitely not intended.

It may look complex when compiled but the [oxide](https://github.com/ephox/oxide) skin project is fairly well structured so I don't imagine these are difficult things to fix.

`HtmlPanel` is an escape hatch, only used by the Help dialog, that we have pondered removing (but we're not sure what we'd do instead). If you don't mind showing us, I'm very curious what your dialog does that requires the label to be different. You can email me if you don't want to share it publicly. If there are use cases out there that v5 makes difficult we would love to know _before_ we release :)

**Comment by Paril** [comment]
We only deal with the compiled versions, that might be where the majority of our issues come from. Our workflow just isn't designed to deal with including TypeScript in here, and having a copy/submodule of tinymce just for editing a few things in the skin is a bit hefty for what we wanna do with it.

I used HtmlPanel to force "whitespace" to wrap so that that label would flow properly, but if that's a bug then that's not a valid use case. I do like the flexibility of HtmlPanel, but at the same time I know it's gonna cause issues since people are going to try to use it to make custom controls and realize that that's not gonna work out for them.

So far I've been able to handle all of the changes that 5.x has thrown at us fairly effectively, and the new dialog system is fairly clean. The lack of flexibility for DOM editing does concern me going forward but I haven't given editing the skin files a proper shot yet. Once I'm finished testing/converting things over to 5.x I'll see what else is left.

**Comment by TheSpyder** [comment]
We might be talking about slightly different things. The TypeScript code is only in the theme which defines the DOM structure and UI behaviour. The version 4 theme is known as `modern`, version 5 is `silver`. The skin is pure CSS, compiled from less source files - version 4 had `lightgrey`, for v5 it's `oxide` and they're in a separate repository :)

**Comment by Paril** [comment]
Oh I see. Yeah, css-wise it's not an issue, we have our own file that is loaded afterwards so we can fix anything we need, but on an individual control basis the DOM/UI behavior is probably what most people are going to be needing custom IDs for, really.

**Comment by Evertvdw** [comment]
Hi there, is there a solution for this? I have the same problem as described in the first post. I want to use a `collection` results field just like the emoticons plugin, but then to show an icon library (Material icons in my case) in the same manner. I can sort of get it to work by putting a font-family on the build in `.tox-collection__item-icon` class but the material icons are not characters as the emoticons are so I require some different height/width values for my custom plugin. But I don't see a way to style this `.tox-collection__item-icon` class only for my plugin. If I change something there in general I also mess with the emoticons plugin, which is what I don't want. 

Long story short, a way for me to put a custom class/id on my own plugin would allow me to override styling only for my own plugin using the built in classes. It still aint pretty but it would work.

**Comment by Evertvdw** [comment]
I did a bit more digging and found some sort of workaround, after `windowManager.open()` use the DomQuery utils to add a custom class to the created .tox-dialog, like this:
```js
var dialogApi = editor.windowManager.open(getInitialState());
tinymce.dom.DomQuery(".tox-dialog").addClass("test-test");
```

**Comment by mikkorantalainen** [comment]
> ```js
> tinymce.dom.DomQuery(".tox-dialog").addClass("test-test");
> ```

Note that hack this would add that class to *all* dialogs if multiple are open. I think TinyMCE should support adding custom class to the dialog so I can apply custom style to specific dialog only. I understand that it doesn't match the ideology that TinyMCE is trying to do here but for one-off style tweaks it avoids creating fully custom UI element just for some visual tweak.
